File: VRTBannerCustomEventVungle.m
//
//  VRTBannerCustomEventGoogleMobileAds.m
//
//  Created by Scott McCoy on 5/9/19.
//  Copyright © 2019 VRTCAL. All rights reserved.
//

//Header
#import "VRTBannerCustomEventVungle.h"

//Dependencies
#import "VRTVungleManager.h"

@interface VRTBannerCustomEventVungle()
@property UIView* containerView;
@end


//Vungle Banner Adapter, Vrtcal as Primary
@implementation VRTBannerCustomEventVungle

- (void) loadBannerAd {
    NSString *placementId = [self.customEventConfig.thirdPartyCustomEventData objectForKey:@"adUnitId"];

    
    if (placementId == nil) {
        VRTError *error = [VRTError errorWithCode:VRTErrorCodeCustomEvent message:@"No placement Id"];
        [self.customEventLoadDelegate customEventFailedToLoadWithError:error];
        return;
    }
    
    CGRect frame = CGRectMake(
        0,
        0,
        self.customEventConfig.adSize.width,
        self.customEventConfig.adSize.height
    );
    
    //Must precede loadPlacementWithID as vungleAdPlayabilityUpdate is often called immediately
    self.containerView = [[UIView alloc] initWithFrame:frame];
    
    NSError *error = [[VRTVungleManager singleton]
        loadPlacementWithID:placementId
        withSize:VungleAdSizeBanner
        vrtVungleManagerDelegate:(id <VRTVungleManagerDelegate>) self
    ];
    
    if (error) {
        [self.customEventLoadDelegate customEventFailedToLoadWithError:error];
        return;
    }
    
    
}

- (UIView*) getView {
    return self.containerView;
}


#pragma mark - VRTVungleManagerDelegate
- (void)vungleAdViewedForPlacement:(nullable NSString *)placementID {
    [self.customEventShowDelegate customEventShown];
}

- (void)vungleDidCloseAdForPlacementID:(nonnull NSString *)placementID {
    [self.customEventShowDelegate customEventDidDismissModal:VRTModalTypeUnknown];
}

- (void)vungleDidShowAdForPlacementID:(nullable NSString *)placementID {
    [self.customEventShowDelegate customEventShown];
}

- (void)vungleRewardUserForPlacementID:(nullable NSString *)placementID {
    // No VRT analog
}

- (void)vungleTrackClickForPlacementID:(nullable NSString *)placementID {
    [self.customEventShowDelegate customEventClicked];
}

- (void)vungleWillCloseAdForPlacementID:(nonnull NSString *)placementID {
    [self.customEventShowDelegate customEventWillDismissModal:VRTModalTypeUnknown];
}

 - (void)vungleWillLeaveApplicationForPlacementID:(nullable NSString *)placementID {
    [self.customEventShowDelegate customEventWillLeaveApplication];
}

- (void)vungleWillShowAdForPlacementID:(nullable NSString *)placementID {
    // No VRT Analog
}

- (void)vungleAdPlayabilityUpdate:(BOOL)isAdPlayable placementID:(nullable NSString *)placementID error:(nullable NSError *)error {
    
    if (error) {
        [self.customEventLoadDelegate customEventFailedToLoadWithError:error];
        return;
    }
    
    if (!isAdPlayable) {
        VRTError *vrtError = [VRTError errorWithCode:VRTErrorCodeNoFill message:@"Vungle Ad Not Playable"];
        [self.customEventLoadDelegate customEventFailedToLoadWithError:vrtError];
        return;
    }
    
    [[VRTVungleManager singleton] showBanner:placementID inContainerView:self.containerView];
    [self.customEventLoadDelegate customEventLoaded];
}
@end

File: VRTBannerCustomEventVungle.swift

//
//  VRTBannerCustomEventGoogleMobileAds.h
//
//  Created by Scott McCoy on 5/9/19.
//  Copyright © 2019 VRTCAL. All rights reserved.
//

//
//  VRTBannerCustomEventGoogleMobileAds.m
//
//  Created by Scott McCoy on 5/9/19.
//  Copyright © 2019 VRTCAL. All rights reserved.
//

//Header
import VrtcalSDK

//Vungle Banner Adapter, Vrtcal as Primary

class VRTBannerCustomEventVungle: VRTAbstractBannerCustomEvent {
    private var containerView: UIView?
    
    override func loadBannerAd() {
        guard let placementId = customEventConfig.thirdPartyCustomEventData["adUnitId"] as? String else {
            let vrtError = VRTError(
                vrtErrorCode: .customEvent,
                message: "No placement Id"
            )
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }
        
        let frame = CGRect(
            x: 0,
            y: 0,
            width: customEventConfig.adSize.width,
            height: customEventConfig.adSize.height)
        
        //Must precede loadPlacementWithID as vungleAdPlayabilityUpdate is often called immediately
        containerView = UIView(frame: frame)
        
        let error = VRTVungleManager.singleton.loadPlacement(
            withID: placementId,
            with: .banner,
            vrtVungleManagerDelegate: self
        )
        
        if let error {
            let vrtError = VRTError(vrtErrorCode: .customEvent, error: error)
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }
    }
    
    override func getView() -> UIView? {
        return containerView
    }
}

// MARK: - VRTVungleManagerDelegate
extension VRTBannerCustomEventVungle: VRTVungleManagerDelegate {

    func vungleAdViewed(forPlacement placementID: String?) {
        customEventShowDelegate?.customEventShown()
    }

    func vungleDidCloseAd(forPlacementID placementID: String) {
        customEventShowDelegate?.customEventDidDismissModal(.unknown)
    }

    func vungleDidShowAd(forPlacementID placementID: String?) {
        customEventShowDelegate?.customEventShown()
    }

    func vungleRewardUser(forPlacementID placementID: String?) {
        // No VRT analog
    }

    func vungleTrackClick(forPlacementID placementID: String?) {
        customEventShowDelegate?.customEventClicked()
    }

    func vungleWillCloseAd(forPlacementID placementID: String) {
        customEventShowDelegate?.customEventWillDismissModal(.unknown)
    }

    func vungleWillLeaveApplication(forPlacementID placementID: String?) {
        customEventShowDelegate?.customEventWillLeaveApplication()
    }

    func vungleWillShowAd(forPlacementID placementID: String?) {
        // No VRT Analog
    }

    func vungleAdPlayabilityUpdate(_ isAdPlayable: Bool, placementID: String?, error: Error?) {

        if let error {
            let vrtError = VRTError(vrtErrorCode: .customEvent, error: error)
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }

        if !isAdPlayable {
            let vrtError = VRTError(vrtErrorCode: .customEvent, message: "Vungle Ad Not Playable")
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }

        guard let containerView else {
            let vrtError = VRTError(vrtErrorCode: .customEvent, message: "Container view nil")
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }
        
        if let error = VRTVungleManager.singleton.showBanner(
            placementId: placementID ?? "",
            inContainerView: containerView
        ) {
            let vrtError = VRTError(vrtErrorCode: .customEvent, error: error)
            customEventLoadDelegate?.customEventFailedToLoad(vrtError: vrtError)
            return
        }

        customEventLoadDelegate?.customEventLoaded()
    }
}
